### PR TITLE
feat(consumer): Support incremental cooperative rebalancing

### PR DESCRIPTION
**Pull request by lynnagara**
KafkaConsumer now supports the `cooperative-sticky` partitioning strategy.
The goal of this change is to try and reduce the amount of unnecessary rebalances
that happen during Kubernetes rolling deployments.

### CODE REVIEW

File: arroyo/backends/kafka/consumer.py [hunk 1]
```diff
@@ -246,40 +252,63 @@ def assignment_callback(
         ) -> None:
             self.__state = KafkaConsumerState.ASSIGNING
 
-            try:
-                assignment: MutableSequence[ConfluentTopicPartition] = []
+            if self.__incremental_cooperative is True:
```
Comment (lynnagara) [inline arroyo/backends/kafka/consumer.py:255]: This diff got pretty weird because of indentation. This block from here to line 273 is new though, and could use a close review. It's quite different to what happens with eager rebalancing (lines 276-305), which is essentially unchanged.

**Review by fpacifici — COMMENTED** [review]
Have you tried to test locally how this performs?
Specifically:
- how long it takes to assign a new partition to a new consumer showing up 
- how long it takes to revoke all remaining partitions (if any) from a consumer that is leaving the group? Does the group wait `scheduled.rebalance.max.delay.ms` before reassigning the partitions to a new consumer?
- how long does it take to have partitions reassigned in case a node leaves and rejoins? 

I think we should do a test locally before doing it in prod.

File: arroyo/backends/kafka/consumer.py [hunk 2]
```diff
@@ -182,6 +183,11 @@ def __init__(
                 "invalid value for 'enable.auto.offset.store' configuration"
             )
 
+        self.__incremental_cooperative = incremental_cooperative
+
+        if self.__incremental_cooperative is True:
+            configuration["partition.assignment.strategy"] = "cooperative-sticky"
```
Comment (fpacifici) [inline arroyo/backends/kafka/consumer.py:189]: Why providing `incremental_cooperative` as a dedicated flag instead of passing it through the `configuration` mapping like all the others config parameters ?

File: arroyo/backends/kafka/consumer.py [hunk 3]
```diff
@@ -431,29 +471,33 @@ def __validate_offsets(self, offsets: Mapping[Partition, int]) -> None:
...
-            self.__consumer.assign(
-                [
-                    ConfluentTopicPartition(
-                        partition.topic.name, partition.index, offset
-                    )
-                    for partition, offset in offsets.items()
-                ]
+        for partition, offset in offsets.items():
+            self.__consumer.seek(
+                ConfluentTopicPartition(partition.topic.name, partition.index, offset)
```
Comment (fpacifici) [inline arroyo/backends/kafka/consumer.py:499]: Why this change ?

File: arroyo/backends/kafka/consumer.py (hunk 2 above)
Comment (lynnagara) [reply arroyo/backends/kafka/consumer.py:189]: I had it the other way originally but figured it was easier to pass a boolean than remember this string then check for it to decide whether to apply incremental_assign. Thinking about it again, I might switch back to avoid changing the interface of this class though.

File: arroyo/backends/kafka/consumer.py [hunk 4]
```diff
@@ -246,40 +252,74 @@ def assignment_callback(
...
-                        for partition, offset in offsets.items()
-                    ]
-                )
+            if self.__incremental_cooperative is True:
+                try:
+                    incremental_assignment: MutableSequence[
+                        ConfluentTopicPartition
+                    ] = []
+
+                    for partition in partitions:
```
Comment (fpacifici) [inline arroyo/backends/kafka/consumer.py:261]: If I am getting this right, you are not asking the broker anymore for the last committed offset. 
Which means that you are invoking `incremental_assign` with the offset you received in the `assignment_callback` call. 
Is this intentional? In the stop-the-world rebalancing, fetching the last committed offset and seeking that was meant to ensure all partitions had the same behavior (whether they were previously assigned to the same consumer or not). All partitions would be reset to the last committed offset.
I would expect it to be impossible, in the incremental case, to receive a partition you were already owning before (unless the consumer restarted). Is this guaranteed or can you still be assigned a partition you already owned before rebalancing?
If it is guaranteed and all partitions you received are "new" to this consumer, are you not resetting the offset because the offset you receive are supposed to already be the last committed ones ?

File: arroyo/backends/kafka/consumer.py (hunk 3 above)
Comment (lynnagara) [reply arroyo/backends/kafka/consumer.py:499]: I found it pretty unintuitive that the same `__seek` implementation was being called from both `seek()` and the assign callback then checking the KafkaConsumerState to effectively determine which one was making the call. Now assigning function is separate and never calls this `__seek` function anymore.

File: arroyo/backends/kafka/consumer.py [hunk 5]
```diff
@@ -246,40 +252,74 @@ def assignment_callback(
...
+                        [
+                            ConfluentTopicPartition(
+                                partition.topic.name, partition.index, offset
+                            )
+                            for partition, offset in offsets.items()
+                        ]
+                    )
+
+                    for partition in offsets:
+                        self.__paused.discard(partition)
```
Comment (fpacifici) [inline arroyo/backends/kafka/consumer.py:319]: Any reason for relying on the `resume` method for the incremental method and, instead, reimplementing its logic here for the stop-the-world rebalancing ?

File: arroyo/backends/kafka/consumer.py (hunk 3 above)
Comment (fpacifici) [reply arroyo/backends/kafka/consumer.py:499]: Also, now this method is only called by the `seek` public method. Why not inlining it there ?

File: tests/backends/test_kafka.py [hunk 6]
```diff
@@ -133,6 +144,70 @@ def test_auto_offset_reset_error(self) -> None:
...
+        {
+            **configuration,
+            "auto.offset.reset": "earliest",
+            "enable.auto.commit": False,
+            "enable.auto.offset.store": False,
+            "group.id": group_id,
+            "session.timeout.ms": 10000,
+        },
+        incremental_cooperative=True,
+    )
```
Comment (fpacifici) [inline tests/backends/test_kafka.py:178]: Thanks.
By the way, what happens if a number of consumers are incremental and others are not ? Which means what will happen at the first deployment ?

File: tests/backends/test_kafka.py [hunk 7]
```diff
@@ -133,6 +144,70 @@ def test_auto_offset_reset_error(self) -> None:
...
+        producer
+    ), closing(consumer_a), closing(consumer_b):
+        for i in range(10):
+            for j in range(partitions_count):
+                producer.produce(
+                    Partition(topic, j),
+                    KafkaPayload(None, f"{j}-{i}".encode("utf8"), []),
+                )
+
+        consumer_a.subscribe([topic])
```
Comment (fpacifici) [inline tests/backends/test_kafka.py:190]: Why not providing callbacks so that you can assert that only one partition is transferred ?

File: arroyo/backends/kafka/consumer.py (hunk 2 above)
Comment (fpacifici) [reply arroyo/backends/kafka/consumer.py:189]: > I might switch back to avoid changing the interface of this class though.

Also we would not reintroduce two separate ways to provide Kafka config that took us long to clean up

File: tests/backends/test_kafka.py (hunk 7 above)
Comment (lynnagara) [reply tests/backends/test_kafka.py:190]: I guess that could've worked too. Still this works well enough in checking the same thing and it's copied from the test of the eager consumer.

File: arroyo/backends/kafka/consumer.py (hunk 5 above)
Comment (lynnagara) [reply arroyo/backends/kafka/consumer.py:319]: No reason, I just mostly avoided touching this implementation (it's just indented differently so it shows up here). But yes it's cleaner to reuse `resume` everywhere, I'll switch it over.

File: arroyo/backends/kafka/consumer.py (hunk 3 above)
Comment (lynnagara) [reply arroyo/backends/kafka/consumer.py]: I was avoiding touching a lot of code earlier, but yes it is better inline since no one else calls it now. Updated.

File: tests/backends/test_kafka.py (hunk 7 above)
Comment (fpacifici) [reply tests/backends/test_kafka.py:190]: Right but there is no way to tell whether all partitions were revoked and then reassigned or whether the incremental assignment actually works.

Comment (lynnagara) [reply tests/backends/test_kafka.py:190]: Yeah this was not about test Kafka consumer internals, just that we end up with one partition assigned to each consumer and consuming on both works ok.

File: arroyo/backends/kafka/consumer.py (hunk 4 above)
Comment (lynnagara) [reply arroyo/backends/kafka/consumer.py:261]: I don't *think* you can receive a partition you already owned, but in any case I don't quite get why we would need to check the offset since the offset you are receiving should already be the correct one. The other issue I noticed with `committed()` is that it hangs forever (or times out if you pass it a timeout) for the cooperative-sticky strategy in the scenario where it's a new consumer group and no offset has been committed yet.

Comment (fpacifici) [reply arroyo/backends/kafka/consumer.py:259]: This will be important to verify. 
The reason for getting the `committed()` offsets is so that all partitions start in a consistent way from the last committed offset after a rebalance (without doing so the partitions that are reassigned to you would start from the last consumed offset). 
This behavior is visible to the application, but if we commit during revoke we should not have any issue as the two behaviors would be identical. 
If we never receive a partition we already owned and we consistently commit on revoke we are good. But if we fail in committing during revoke and we get the same partition we owned in the past I don't know what would happen if we do not reset the offsets. 
If `committed()` is taking much longer, then it is probably a non starter. Could you please verify that even if we did receive the same partition again on subscriptions (where we are testing this), we will not risk missing messages ?

Comment (lynnagara) [reply arroyo/backends/kafka/consumer.py:259]: > The reason for getting the `committed()` offsets is so that all partitions start in a consistent way from the last committed offset after a rebalance (without doing so the partitions that are reassigned to you would start from the last consumed offset).

I don't think this is true. The offsets passed to the assignment callback are already the committed ones not just the ones that are consumed.

Kafka guarantees that
- only new partitions are passed in this callback not previously owned ones
- The revoke callback is always triggered before the assignment one.

The reason for committing offsets during revoke is to avoid double processing them so the new consumer doesn't get that same offset again.

**Comment by lynnagara** [comment]
> * how long it takes to assign a new partition to a new consumer showing up
> * how long it takes to revoke all remaining partitions (if any) from a consumer that is leaving the group? Does the group wait `scheduled.rebalance.max.delay.ms` before reassigning the partitions to a new consumer?
> * how long does it take to have partitions reassigned in case a node leaves and rejoins?

It's typically not as long as `scheduled.rebalance.max.delay.ms` which is 5 minutes by default. All of these usually take around 3 seconds in my local testing with 2 consumers. The test case I added adds a new consumer to an existing group and waits for a partition to assigned to it. We wait for up to 10 seconds max and it generally always passes.

File: arroyo/backends/kafka/consumer.py (hunk 4 above)
Comment (fpacifici) [reply arroyo/backends/kafka/consumer.py:259]: > The offsets passed to the assignment callback are already the committed ones not just the ones that are consumed.

Are you sure. I thought this was the reason for making this consumer behave differently from the standard Kafka consumer https://github.com/getsentry/arroyo/blob/main/arroyo/backends/kafka/consumer.py#L109-L118.

Anyway, if Kafka guarantees that only new partitions are passed and not previously owned one, we should be good.

File: tests/backends/test_kafka.py (hunk 6 above)
Comment (fpacifici) [reply tests/backends/test_kafka.py]: Ok, let's do it one day I am on PTO. 
Out of curiosity, what happens if you start a consumer that works with the incremental rebalancing system to a group that is using the standard system ?

File: arroyo/backends/kafka/consumer.py (hunk 2 above)
Comment (fpacifici) [reply arroyo/backends/kafka/consumer.py]: I am talking about all those fields we were passing via CLI that we are slowly moving to settings.

**Review by fpacifici — APPROVED** [review]
Seems good to try

**Comment by fpacifici** [comment]
I think you should update the docstring of the KafkaConsumer class so that it reflects the behavior on incremental cooperative rebalancing. Specifically this https://github.com/getsentry/arroyo/blob/main/arroyo/backends/kafka/consumer.py#L109-L118

File: arroyo/backends/kafka/consumer.py (hunk 4 above)
Comment (lynnagara) [reply arroyo/backends/kafka/consumer.py:259]: Yes, but this only applies to the partitions that are being newly assigned to the consumer. Previously all of the partitions (including those already assigned to the consumer) would be provided in the callback, now only the incremental ones are. So we are no longer rewinding back those partitions which the consumer is maintaining across a rebalance like we use to, those just continue from the same place.

File: arroyo/backends/kafka/consumer.py (hunk 2 above)
Comment (lynnagara) [reply arroyo/backends/kafka/consumer.py]: Unfortunately passing via CLI is still the easiest option currently as we don't have a mechanism yet to provide different settings for consumers/producers of a topic via settings.

**Comment by mwarkentin** [comment]
@lynnagara 🎉 

Let me know if / when you need some help testing this. :)
